src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -174,11 +174,15 @@ SymbolKind.NetModule or
                     {
                         // We may still have a cached generator; we'll have to remember to run generators again since we are making some
                         // change here. We'll also need to update the other state of the driver if appropriate.
-                        var generatorInfo = state.GeneratorInfo.WithDocumentsAreFinal(false);
+
+                        // The no compilation state can never be in the 'DocumentsAreFinal' state.  The only place where
+                        // we start with the NoCompilationState is the 'Empty' instance (where DocumentsAreFinal=false).
+                        // And then this is the only place where we get a NoCompilationState and create a new instance.
+                        // So there is no way to ever transition this to the DocumentsAreFinal=true state.
+                        Contract.ThrowIfTrue(state.GeneratorInfo.DocumentsAreFinal);
```

Comment:
I'd almost say move this comment to the assert in the NoCompilationState constructor, and then you can maybe drop this extra assert here.

I think the statement is now correct: when we had weak compilations, if the compilation fell away and the solution was forked, we wouldn't have any compilation at all, so we wouldn't create an inProgressState because there was nothing to build off of or transform. But: the generated result was still stale.

Reply:
sounds good to me.